src/components/TodoItem.tsx

```diff
@@ -0,0 +1,247 @@
+      .then(updatedTodo => {
+        setTodos(prevTodos => {
+          if (prevTodos) {
+            return prevTodos.map(currentTodo => {
+              if (currentTodo.id === todo.id) {
+                return updatedTodo;
+              }
+
+              return currentTodo;
+            });
+          }
+
+          return null;
+        });
+      })
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        setTodosInProcess(prevTodosIds => (
+          prevTodosIds.filter(id => todoId !== id)
+        ));
+
+        setIsEditing(false);
+      });
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setTodosInProcess(prevTodosIds => [...prevTodosIds, todoId]);
+
+    return postService.removeTodo(todoId)
+      .then(() => {
+        setTodos(prevTodos => {
+          if (prevTodos) {
+            return prevTodos.filter(currentTodo => currentTodo.id !== todoId);
+          }
+
+          return null;
+        });
+      })
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() => {
+        setTodosInProcess(prevTodosIds => (
+          prevTodosIds.filter(id => todoId !== id)
+        ));
+
+        setIsEditing(false);
+      });
+  };
+
+  const reset = () => {
+    setNewTitle(todo.title);
+    setIsEditing(false);
+  };
+
+  const onEsc = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      reset();
+    }
+  };
+
+  const handleFormSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (todo.title === newTitle) {
+      setIsEditing(false);
+
+      return;
+    }
+
+    if (!newTitle.length) {
+      deleteTodo(todo.id)
+        .finally(() => setIsEditing(false));
+    }
+
+    if (newTitle.length) {
+      updateTitle(todo.id, { title: newTitle });
+    }
+  };
+
+  const handleOnBlur = () => {
+    if (todo.title === newTitle) {
+      setIsEditing(false);
+
+      return;
+    }
+
+    if (!newTitle.length) {
+      deleteTodo(todo.id)
+        .finally(() => setIsEditing(false));
+    }
+
+    if (newTitle.length) {
+      updateTitle(todo.id, { title: newTitle.trim() })
+        .finally(() => setIsEditing(false));
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn(
+        'todo',
+        {
+          completed: todo.completed,
+        },
+      )}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onClick={() => {
+            updateCompleted(todo.id, { completed: !todo.completed });
+          }}
+        />
+      </label>
+
+      {!isEditing ? (
+        <>
+          <span
+            data-cy="TodoTitle"
+            className="todo__title"
+            onDoubleClick={() => {
+              setNewTitle(newTitle.trim());
+              setIsEditing(true);
+            }}
```

Comment:
Pls move the function inside `onDoubleClick` handler to a separate const and use as a callback here